### DISCUSSION TITLE
Good pricing with conversation architecture in Sequelize

### DISCUSSION THREAD

**Discussion by doneumark** [sequelize/sequelize#14721] [Help & Questions]
Hi!
How do you handle price with currency in Sequelize?
If you have a price field inside a Product model, do you save it as a json with the currency and the amount inside?
```
Product {
    ...
    somePriceField: { currency: STRING, amount: DECIMAL }
}
```

Or do you save the currency and the amount on two different fields?
```
Item {
    ...
    somePriceFieldCurrency: STRING,
    somePriceFieldAmount: DECIMAL
}
```

Or do you maybe reference a Price model, and JOIN it to the Product Model?

Also - how do you handle currency conversation, where the price and currency should be calculated accordingly to another **different** currency?

Do you use a hooks (afterFind, beforeFind)?
Static function (Product.findAndConvert('EUR', options))?
Scope (Product.scope('EUR'))?

Remember you need to tell what to convert **to**, which can be an async question.

I would love a good example.

**Comment by ephys** [comment]
It's up to you really, my only note is do not use float, number, or double (IEE 754) to store the amount.

If you're using a JSON field, store it as a string
If you're storing it as its own column, use `DataTypes.DECIMAL` for it

**Reply by doneumark** [reply]
I wrote FLOAT only to make thing look simpler, turns out its too much of a red flag to people whenever they read it :)
Of course the amount field is save as DECIMAL. Edited the post.

**Reply by ephys** [reply]
The rest really depends on you. Storing it as a JSON object with price & currency separated is a good solution. Splitting in two columns is also good. For the convenience of being able to pass them around together I would personally opt for the JSON version. I don't see a reason to do it as part of a separate model, so no JOIN imo.

For currency conversion: that part is much trickier. IMO I would not convert automatically. I would either:
- Allow specifying the amount in different currencies, manually
- Give the amount in EUR to the payment provider and let them handle the conversion at the current exchange rate

If you really need to convert automatically, there are APIs that will give you the current exchange rates. I don't think this concern belongs in your Sequelize models though, it belongs somewhere in your business logic

**Reply by doneumark** [reply]
Because the model is a very "used" model in the app, and also its not the ONLY model who uses prices, I want a scalable solution that will prevent all of the code duplication my code has regarding the conversation.
Thus, i'm wondering WHERE is the best place to execute the conversation, and would it require saving the currency rates in a table and use joins? calling an async function AFTER the query? or maybe BEFORE? or just saving all of the the different currency values beforehand?
I would just love to hear examples from people who tackled this situation. Ofcourse there are many ways to overcome this.
